### PR TITLE
`Locale` - storages concept

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #85, #83

An attempt to implement storages concept suggested in https://github.com/yiisoft/yii-middleware/pull/82#issuecomment-1552720246.

TODO:

- [ ] Handle the case of reading from storage, but not saving to it.

### CODE REVIEW

File: src/Storage/LocaleStorageWithResponseInterface.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Middleware\Storage;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+interface LocaleStorageWithResponseInterface extends LocaleStorageInterface
```
Comment (samdark) [inline src/Storage/LocaleStorageWithResponseInterface.php:10]: It has both request and response but called `WithResponse`.

File: src/Locale.php [hunk 2]
```diff
@@ -37,16 +36,17 @@ final class Locale implements MiddlewareInterface
...
-    private string $sessionName = self::DEFAULT_LOCALE_NAME;
-    private string $cookieName = self::DEFAULT_LOCALE_NAME;
     /**
      * @psalm-var array<string, string>
      */
     private array $supportedLocales;
+    /**
+     * @var LocaleStorageInterface[]
+     */
+    private array $storages;
```
Comment (samdark) [inline src/Locale.php:49]: Could be done as the following:

```php
interface LocaleStorageInterface {}

class LocaleStorage implements LocaleStorageInterface
{
    public function __construct(
        private array $localeStorages
    )
    {}

    public function save()
    {
        foreach ($this->storages as $storage)
        {
            $storage->save();
        }
    }
}

class CookieStorage implements LocaleStorageInterface {}
class SessionStorage implements SessionStorageInterface {}
```

Comment (arogachev) [reply src/Locale.php:49]: How to combine it with logger?

File: src/Locale.php [hunk 3]
```diff
@@ -83,6 +74,14 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
             return $handler->handle($request);
         }
 
+        foreach ($this->storages as $storage) {
+            if (!$storage instanceof LocaleStorageInterface) {
+                $interfaceName = LocaleStorageInterface::class;
+
+                throw new InvalidArgumentException("Locale storage must implement \"$interfaceName\".");
```
Comment (xepozz) [inline src/Locale.php:81]: Specify what the storage exactly

File: src/Event/SetLocaleEvent.php [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 namespace Yiisoft\Yii\Middleware\Event;
 
-final class LocaleEvent
+final class SetLocaleEvent
```
Comment (xepozz) [inline src/Event/SetLocaleEvent.php:7]: LocaleSet / AfterLocaleSet?

File: src/Locale.php [hunk 5]
```diff
@@ -99,9 +98,18 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
...
-                $cookieParameters = $request->getCookieParams();
-                $locale = $this->getLocaleFromCookies($cookieParameters);
+                foreach ($this->storages as $storage) {
+                    if ($storage instanceof LocaleStorageWithHttpFlowInterface) {
+                        $storage = $storage->withRequest($request);
+                    }
+
+                    $locale = $storage->get();
+                    if ($locale !== null) {
+                        $this->logger->debug("Locale '$locale' found in cookie.");
```
Comment (xepozz) [inline src/Locale.php:108]: Use logger params instead

File: src/Locale.php [hunk 6]
```diff
@@ -130,11 +138,21 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
...
 
-        if ($this->saveLocale) {
-            $response = $this->saveLocale($locale, $response);
+        foreach ($this->storages as $storage) {
+            if ($storage instanceof LocaleStorageWithHttpFlowInterface) {
+                $storage = $storage->withResponse($response);
+            }
+
+            $storageName = $storage->getName();
+            $this->logger->debug("Saving found locale to $storageName.");
```
Comment (xepozz) [inline src/Locale.php:150]: Same

File: src/Storage/LocaleStorageWithHttpFlowInterface.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Middleware\Storage;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+interface LocaleStorageWithHttpFlowInterface extends LocaleStorageInterface
```
Comment (xepozz) [inline src/Storage/LocaleStorageWithHttpFlowInterface.php:10]: HttpLocaleStorage?

File: src/Storage/CookieLocaleStorage.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        $this->cookie = $this->cookie->withValue($value);
+
+        if ($this->response !== null) {
+            $this->response = $this->cookie->addToResponse($this->response);
+        }
+    }
+
+    public function get(): ?string
+    {
+        return $this?->request?->getCookieParams()[$this->cookie->getName()] ?? null;
```
Comment (xepozz) [inline src/Storage/CookieLocaleStorage.php:37]:
```suggestion
        return $this->request?->getCookieParams()[$this->cookie->getName()] ?? null;
```

File: src/Storage/CookieLocaleStorage.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    public function getName(): string
+    {
+        return 'cookie';
+    }
+
+    public function set(string $value): void
+    {
+        $this->cookie = $this->cookie->withValue($value);
+
+        if ($this->response !== null) {
```
Comment (xepozz) [inline src/Storage/CookieLocaleStorage.php:30]: Looks like performance degradation. What to you think about "flush" stage instead?

File: src/Storage/SessionLocaleStorage.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public function set(string $value): void
+    {
+        $this->session->set($this->itemName, $value);
+    }
+
+    public function get(): ?string
+    {
+        return null;
```
Comment (xepozz) [inline src/Storage/SessionLocaleStorage.php:29]: ?

**Pull request closed without merging**